### PR TITLE
Refactor window storage

### LINKED ISSUE #621: [ENHANCEMENT]: Improve `make_window_extent` public interface
### Is your feature request related to a problem? Please describe.

We currently recommend users construct the window extent type based on the ref_type (or container type), as shown here: https://github.com/NVIDIA/cuCollections/blob/ae3ac6e40a3dd0c48d32baa3162a93635efb3c2a/examples/static_set/device_subsets_example.cu#L146

However, this approach is confusing because the actual reference type is determined by its members and after obtaining the window extent, the actual reference type differs from the one initially used to compute the window extent. This is particularly problematic when adding explicit instantiations, as users often struggle to determine the correct type.

### Describe the solution you'd like

Add new overloads of `make_window_extent`, e.g.
```cpp
template <typename ProbingScheme, typename Storage, typename Extent>
[[nodiscard]] auto constexpr make_window_extent(ProbingScheme probing_scheme, Storage storage, Extent extent) {
   ...
}
```

Deprecate the old APIs taking window size and CG size as template parameters:
```cpp
template <int32_t CGSize, int32_t WindowSize, typename SizeType, std::size_t N>
[[nodiscard]] auto constexpr make_window_extent(extent<SizeType, N> ext);


template <int32_t CGSize, int32_t WindowSize, typename SizeType>
[[nodiscard]] auto constexpr make_window_extent(SizeType size);
```

Move the below APIs taking container as input to detail namespace since it should be only used when constructing ref inside the container:
```cpp
template <typename Container, typename SizeType, std::size_t N>
[[nodiscard]] auto constexpr make_window_extent(extent<SizeType, N> ext);

template <typename Container, typename SizeType>
[[nodiscard]] auto constexpr make_window_extent(SizeType size);
```

### PR DESCRIPTION
**Pull request by PointKernel**
Closes #621

Based on the offline discussions, this PR replaces the `window` logic with `bucket` and adds new overloads of `make_bucket_extent` so OA no longer relies on the `Container` type to determine the bucket extent.

### CODE REVIEW

File: include/cuco/bucket_storage.cuh [hunk 1]
```diff
@@ -29,200 +29,210 @@
...
-/// Window type alias
-template <typename T, int32_t WindowSize>
-using window = detail::window<T, WindowSize>;
-
-/// forward declaration
-template <typename T, int32_t WindowSize, typename Extent>
-class aow_storage_ref;
+/// Alias for bucket
+template <typename T, int32_t BucketSize>
+using window = bucket<T, BucketSize>;
```
Comment (sleeepyjack) [inline include/cuco/bucket_storage.cuh:38]: Do we want to keep this alias for now so we don't break in a gazillion places in cudf?

File: include/cuco/bucket_storage.cuh [hunk 2]
```diff
@@ -29,200 +29,210 @@
...
 
 /**
- * @brief Array of Window open addressing storage class.
+ * @brief Non-owning AoW storage reference type.
```
Comment (sleeepyjack) [inline include/cuco/bucket_storage.cuh]: AoS?

File: include/cuco/bucket_storage.cuh [hunk 3]
```diff
@@ -29,200 +29,210 @@
...
    *
-   * @return Reference of window storage
+   * @return Pointer to the first bucket
    */
-  [[nodiscard]] constexpr ref_type ref() const noexcept;
+  [[nodiscard]] __device__ constexpr bucket_type* data() noexcept;
 
   /**
-   * @brief Initializes each slot in the AoW storage to contain `key`.
+   * @brief Gets buckets array.
```
Comment (sleeepyjack) [inline include/cuco/bucket_storage.cuh]:
```suggestion
   * @brief Gets bucket array.
```

unsure about this

File: include/cuco/bucket_storage.cuh [hunk 4]
```diff
@@ -29,200 +29,210 @@
...
  private:
-  allocator_type allocator_;            ///< Allocator used to (de)allocate windows
-  window_deleter_type window_deleter_;  ///< Custom windows deleter
-  std::unique_ptr<window_type, window_deleter_type> windows_;  ///< Pointer to AoW storage
+  bucket_type* buckets_;  ///< Pointer to the buckets array
 };
 
 /**
- * @brief Non-owning AoW storage reference type.
+ * @brief Array of Bucket open addressing storage class.
```
Comment (sleeepyjack) [inline include/cuco/bucket_storage.cuh]:
```suggestion
 * @brief Array of Buckets open addressing storage class.
```

File: include/cuco/bucket_storage.cuh [hunk 5]
```diff
@@ -29,200 +29,210 @@
...
+  /// Type of the allocator to (de)allocate buckets
+  using allocator_type =
+    typename std::allocator_traits<Allocator>::template rebind_alloc<bucket_type>;
+  using bucket_deleter_type =
+    detail::custom_deleter<size_type, allocator_type>;  ///< Type of bucket deleter
+  using ref_type = bucket_storage_ref<value_type, bucket_size, extent_type>;  ///< Storage ref type
 
   /**
-   * @brief Constructor of AoS storage ref.
+   * @brief Constructor of AoW storage.
```
Comment (sleeepyjack) [inline include/cuco/bucket_storage.cuh]: AoS or Array of buckets?

File: include/cuco/bucket_storage.cuh [hunk 6]
```diff
@@ -29,200 +29,210 @@
...
+  void initialize(value_type key, cuda::stream_ref stream = {});
 
   /**
-   * @brief Returns an array of slots (or a window) for a given index.
+   * @brief Asynchronously initializes each slot in the AoW storage to contain `key`.
    *
-   * @param index Index of the window
-   * @return An array of slots
+   * @param key Key to which all keys in `slots` are initialized
+   * @param stream Stream used for executing the kernel
```
Comment (sleeepyjack) [inline include/cuco/bucket_storage.cuh]: These diffs don't correspond to each other. Dear git, wtf.

File: include/cuco/extent.cuh [hunk 7]
```diff
@@ -125,43 +150,100 @@ template <typename Container, typename SizeType, std::size_t N>
...
+ *
+ * @param ext The input extent
+ *
+ * @throw If the input extent is invalid
+ *
+ * @return Resulting valid `bucket extent`
  */
-template <int32_t CGSize, int32_t WindowSize, typename SizeType, std::size_t N>
+template <typename Container, typename SizeType, std::size_t N>
 [[nodiscard]] auto constexpr make_window_extent(extent<SizeType, N> ext);
```
Comment (sleeepyjack) [inline include/cuco/extent.cuh:218]: Do we keep this for now for backwards compatibility in cudf?

**Review by sleeepyjack — COMMENTED** [review]
Wow, so many changes! I went over most of it but at some point during reading my brain started to come up with every possible pronunciation of the word "bucket". 😵

File: include/cuco/bucket_storage.cuh (hunk 1 above)
Comment (PointKernel) [reply include/cuco/bucket_storage.cuh:38]: exactly

File: include/cuco/extent.cuh (hunk 7 above)
Comment (PointKernel) [reply include/cuco/extent.cuh:218]: yes

**Review by sleeepyjack — APPROVED** [review]

**Pull request merged**